### ISSUE TITLE
campers hardcoding solutions for final projects

### ISSUE THREAD

**Issue by ilenia-magoni** [freeCodeCamp/freeCodeCamp#54063]
The JavaScript final projects have an issue, I don't know if how they are written, there are really specific user stories and no general one. Anyway, it's scarely common that campers try hardcoding the solutions. Should we try to write tests that discourage this?

Or, alternatively, more generic userstories that explain the thing better?
Like, we have
> When the #search-input element contains the value Red and the #search-button element is clicked, an alert should appear with the text "Pokémon not found"

But there isn't a more generic user story that is explicit on the requirement.
> When the #search-input element contains a value that is not a Pokémon name...

**Comment by gikf** [comment]
One option could be trying to join the two, when that's possible. Taking the same example as above:
- One specific user story, using in test the exact _values_ as in the user story.
> When the #search-input element contains the value Red and the #search-button element is clicked, an alert should appear with the text "Pokémon not found"
- One generic user story, where test should be essentially the same as in the specific case, but using different value to search.
> When searched text is not a Pokémon name, an alert should appear with the text "Pokémon not found"

**Comment by ilenia-magoni** [comment]
note, it's not only an issue for the Pokémon search app, I have seen this for the palindrome checker and the roman numeral converter too

**Comment by jeremylt** [comment]
Yeah, campers seem to be struggling really hard with how to create original functions that meet the user story outcomes here.

This isn't wholly surprising, as up to the point of the projects they no longer encounter places where we ask them to generate functions to accomplish a task. I think the fix is twofold. One, we need to test for and prohibit testing to the solutions. Second, we need to give campers smaller opportunities to build single functions from scratch that meet a tested objective - maybe something that looks like https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/#basic-javascript :upside_down_face:

**Comment by jdwilkin4** [comment]
@a2937 and @jeremylt 

> Yeah something like that basic JavaScript path would be great before we get to the html/css/javascript projects considering their bulky size.

that is why the new intro project is being created

https://github.com/freeCodeCamp/CurriculumExpansion/pull/374

**Comment by jdwilkin4** [comment]
I vote for this approach

> One option could be trying to join the two, when that's possible. 

A combination of generic and specific user stories would be ideal

**Comment by jeremylt** [comment]
That's a good project but it's one project though. I'm proposing that we give back the campers the opportunities to do many small one off functions to practice open ended problem solving before they get to the certification projects.

**Comment by jdwilkin4** [comment]
> I'm proposing that we give back the campers the opportunities to do many small one off functions to practice open ended problem solving before they get to the certification projects.

Yeah, I see where you are coming from.

A new GitHub issue would have to be opened for this particular discussion and the rest of the JS team would need to weigh in 👍

**Comment by gikf** [comment]
_Generic_ tests could also add some randomness to the equation. Either by picking test case from a specific set test cases/results, or a more generated test case. For example:

Pokémon Search App:
- Test picks from few valid Pokémon names or IDs, which will be used in test.
- Test checking invalid name or ID could use randomly generated value.

Roman Numeral Converter:
- Randomizing number below `0`, which is used to test the appropriate behavior.

Palindrome Checker:
- Generating strings that fulfills _generic test_ requirements, for example: generating random lowercase word (just letters).

**Comment by ilenia-magoni** [comment]
that last one, it can be also used to test palindromes: generate 4 letters, reverse the order and concat the two together: voilá a palindrome

**Comment by gikf** [comment]
Exactly, and with minimal care, the risk of generating confusing inputs (ie. during one try with generated input `x` test passes, during another with generated input `y` it doesn't) is marginal.

There's one more thing that occurred to me. When _User Stories_ are taken literally, it might be hard to say that hardcoded solution is not fulfilling them. (Leaving out the obvious fact that project is not functional at all). Perhaps user stories should additionally be changed to have only the general versions, while leaving detailed examples to tests? Or present test cases more explicitly as examples in user stories ie.:

> Checker should handle correctly single letter strings. For example: When the #text-input element only contains the letter A and the #check-btn element is clicked, the #result element should contain the text "A is a palindrome"

**Comment by jeremylt** [comment]
I like the idea of language that explains both the goal and the test. Just gotta be careful to phrase it clearly and simply but that would focus first on the target functionality

**Comment by ojeytonwilliams** [comment]
@a2937 @gikf do the linked PRs do enough to close this?

**Comment by c0d1ng-ma5ter** [comment]
Hello. I just found that my PR has a minor bug, so this is still technically not solved. I just submitted a PR to fix those bugs though.
